proxylib/binarymemcached/parser.go

```diff
@@ -0,0 +1,341 @@
+			case "keyExact":
+				br.keyExact = v
+			case "keyPrefix":
+				br.keyPrefix = v
+			case "keyRegex":
+				br.keyRegex = v
+				br.regex = regexp.MustCompile(v)
+			default:
+				panic(fmt.Errorf("Unsupported key: %s", k))
+			}
+		}
+		log.Infof("Parsed BinaryMemcacheRule pair: %v", br)
+		rules = append(rules, &br)
+	}
+	return rules
+}
+
+// BinaryMemcacheParserFactory implements proxylib.ParserFactory
+type BinaryMemcacheParserFactory struct{}
+
+// Create creates binary memcached parser
+func (p *BinaryMemcacheParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("BinaryMemcacheParserFactory: Create: %v", connection)
+	return &BinaryMemcacheParser{connection: connection, injectQueue: make([]queuedInject, 0)}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &BinaryMemcacheParserFactory{}
+
+var binaryMemcacheParserFactory *BinaryMemcacheParserFactory
+
+const (
+	parserName = "binarymemcache"
+)
+
+func init() {
+	log.Info("init(): Registering binaryMemcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, binaryMemcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7BinaryMemcacheRuleParser)
+}
+
+// BinaryMemcacheParser implements proxylib.Parser
+type BinaryMemcacheParser struct {
+	connection *proxylib.Connection
+
+	requestCount uint32
+	replyCount   uint32
+	injectQueue  []queuedInject
+}
+
+var _ proxylib.Parser = &BinaryMemcacheParser{}
+
+const headerSize = 24
+
+// OnData parses binary memcached data
+func (p *BinaryMemcacheParser) OnData(reply, endStream bool, strData []string, offset uint32) (proxylib.OpType, uint32) {
+	log.Infof("OnData with offset %d", offset)
+
+	//TODO: optimise this
+	data := ([]byte(strings.Join(strData, "")))[offset:]
+	log.Infof("Data length: %d", len(data))
+
+	headerMissing := headerSize - len(data)
+
+	if headerMissing > 0 {
+		log.Infof("Did not receive needed header data, need %d more bytes", headerMissing)
+		return proxylib.MORE, uint32(headerMissing)
+	}
+
+	bodyLength := binary.BigEndian.Uint32(data[8:12])
+
+	keyLength := uint32(binary.BigEndian.Uint16(data[2:4]))
+	extrasLength := uint32(data[4])
+
+	if keyLength > 0 {
+		keyMissing := int(headerSize+keyLength+extrasLength) - len(data)
+		if keyMissing > 0 {
+			log.Infof("Did not receive enough bytes for key, need %d more bytes", keyMissing)
+			return proxylib.MORE, uint32(keyMissing)
+		}
+	}
+
+	opcode, key, err := p.parse(data, extrasLength, keyLength)
+	if err != 0 {
+		return proxylib.ERROR, uint32(err)
+	}
+
+	logEntry := &cilium.LogEntry_GenericL7{
+		&cilium.L7LogEntry{
+			Proto: "binarymemcached",
+			Fields: map[string]string{
+				"opcode": strconv.Itoa(int(opcode)),
+				"key":    key,
+			},
+		},
+	}
+
+	// we don't filter reply traffic
+	if reply {
+		log.Infof("reply, passing %d bytes", uint32(len(data)))
+		p.connection.Log(cilium.EntryType_Response, logEntry)
+		p.replyCount++
+		defer p.checkInjectQueue()
```

Comment:
Here the intent is to inject a message in the current (reply) direction after the passed reply. I know that this is currently not documented anywhere, but while injects to the opposite direction do not require an 'operation', injects to the current direction do, otherwise the injected data will just be erased.

I have wanted to keep the Parser API simple, returning one operation at at time. But here we need to first return the pass and then the potential inject from the queue.

To get that effect you should check the inject queue in the very beginning *before* consuming any of the data if `reply` is `true`. That way the checking is only done in the reply direction and while the parser is still at a frame boundary.  You should check the inject queue, and in addition to placing a message to the buffer, return an explicit `proxylib.INJECT` with the correct length. You should do this check even if there is no data available.

If there is no data to inject and there is no data you can return `proxylib.NOP` with zero length instead of MORE with the full header length.

The library currently explicitly calls in to the parser after all received data has been exhausted to give the parser the opportunity to inject after the available input. Once we get this case done we should consider refactoring the reply queueing to the proxylib/Connection, if possible, so that all parsers would not need to re-implement this...